src/Covenant/Type.hs

```diff
@@ -588,3 +656,136 @@ returnHelper xs = case NonEmpty.uncons xs of
     if Vector.length ys == 0
       then pure y
       else Nothing
+
+unify ::
+  ValT Renamed ->
+  ValT Renamed ->
+  Either TypeAppError (Map (Index "tyvar") (ValT Renamed))
+unify expected actual =
+  catchError
+    ( case expected of
+        Abstraction t1 -> case t1 of
+          -- Unifiables unify with everything, and require a substitutional rewrite.
+          Unifiable index1 -> pure . Map.singleton index1 $ actual
+          Rigid level1 index1 -> expectRigid level1 index1
+          Wildcard scopeId1 index1 -> expectWildcard scopeId1 index1
+        ThunkT t1 -> expectThunk t1
+        BuiltinFlat t1 -> expectFlatBuiltin t1
+        BuiltinNested t1 -> case t1 of
+          ListT _ t1' -> expectListOf t1'
+          PairT t11 t12 -> expectPairOf t11 t12
+    )
+    (promoteUnificationError expected actual)
+  where
+    unificationError :: forall (a :: Type). Either TypeAppError a
+    unificationError = Left . DoesNotUnify expected $ actual
+    noSubUnify :: forall (k :: Type) (a :: Type). Either TypeAppError (Map k a)
+    noSubUnify = pure Map.empty
+    expectRigid ::
+      Int -> Index "tyvar" -> Either TypeAppError (Map (Index "tyvar") (ValT Renamed))
+    -- Rigids behave identically to concrete types: they can unify with
+    -- themselves, or any other abstraction, but nothing else. No substitutional
+    -- rewrites are needed.
+    expectRigid level1 index1 = case actual of
+      Abstraction (Rigid level2 index2) ->
+        if level1 == level2 && index1 == index2
+          then noSubUnify
+          else unificationError
+      Abstraction _ -> noSubUnify
+      _ -> unificationError
+    expectWildcard ::
+      Word64 -> Index "tyvar" -> Either TypeAppError (Map (Index "tyvar") (ValT Renamed))
+    -- Wildcards can unify with unifiables, as well as themselves, but nothing
+    -- else. No substitutional rewrites are needed.
+    expectWildcard scopeId1 index1 = case actual of
+      Abstraction (Unifiable _) -> noSubUnify
+      Abstraction (Wildcard scopeId2 index2) ->
+        if scopeId1 /= scopeId2 || index1 == index2
```

Comment:
I believe this condition should be `scopeId1 == scopeId2 && index1 == index2`, i.e. both index and scope have to be the same to unify.

Reply:
Not quite. The rule for wildcards is that they cannot unify with a _different_ wildcard in the _same_ scope. To see why this is, consider `(forall a b . a -> b -> !Integer) -> (forall c . c -> !Integer) -> Integer -> !Integer`. It is possible for `a` and `c` to unify, but not `a` and `b`.

The rule you give is correct for _rigids_, because rigids can only unify with themselves.

Reply:
Ah, I see, so either the scope doesn't match or when it does, the index has to match. 

Okay, that sounds right now with the additional explanation :+1: